

<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>count-reads-fw &mdash; HATCHet 2.1.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
      <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js?v=0218eca2"></script>
      <script src="_static/doctools.js?v=888ff710"></script>
      <script src="_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            HATCHet
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="README.html">HATCHet</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="modules.html">Modules</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">HATCHet</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">count-reads-fw</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/doc_count_reads_fw.md.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="count-reads-fw">
<h1>count-reads-fw<a class="headerlink" href="#count-reads-fw" title="Permalink to this heading">¶</a></h1>
<p>NOTE: This function (formerly called <code class="docutils literal notranslate"><span class="pre">comBBo</span></code>) uses the legacy fixed-width binning described in the HATCHet paper. We recommend using <a class="reference internal" href="doc_count_reads.html"><span class="doc">count-reads</span></a> and <a class="reference external" href="doc_combine-counts"><code class="docutils literal notranslate"><span class="pre">combine-counts</span></code></a> which apply an adaptive binning scheme to ensure that each genomic bin has comparable BAF signal.</p>
<p>This step of HATCHet splits the human reference genome into fixed-width bins (i.e., small genomic regions), and computes the number of sequencing reads aligned to each bin from every given tumor samples and from the matched normal sample.</p>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this heading">¶</a></h2>
<p>count-reads-fw takes in input sorted and indexed BAM files for multiple tumor samples from the same patient, a sorted and index BAM file from a matched-normal sample, and a indexed human reference genome.</p>
<table border="1" class="docutils">
<thead>
<tr>
<th>Name</th>
<th>Description</th>
<th>Usage</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>-T</code>, <code>--tumors</code></td>
<td>A white-space separated list of sorted-indexed BAM files for tumor samples</td>
<td>The tumor samples from the same patient that are jointly analyzed by HATCHet</td>
</tr>
<tr>
<td><code>-N</code>, <code>--normal</code></td>
<td>A sorted-indexed BAM file for matched-normal sample</td>
<td>The matched normal sample for the same patient</td>
</tr>
<tr>
<td><code>-r</code>, <code>--reference</code></td>
<td>A FASTA file</td>
<td>The human reference genome used for germline variant calling</td>
</tr>
</tbody>
</table></section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this heading">¶</a></h2>
<p>count-reads-fw produces three tab-separated files: the first contains the read counts for every genomic bin in every tumor sample (raw counts not normalized), the second contains the read counts for every genomic bin the matched-normal sample (raw counts not normalized), and the third contains a list of the genomic positions that have been identified as germline heterozygous SNPs in the matched-normal sample.</p>
<table border="1" class="docutils">
<thead>
<tr>
<th>Name</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>-O</code>, <code>--outputnormal</code></td>
<td>The output file for the read counts from matched-normal sample</td>
</tr>
<tr>
<td><code>-o</code>, <code>--outputtumors</code></td>
<td>The output file for the read counts from the tumor samples</td>
</tr>
</tbody>
</table></section>
<section id="main-parameters">
<h2>Main parameters<a class="headerlink" href="#main-parameters" title="Permalink to this heading">¶</a></h2>
<p>count-reads-fw</p>
</section>
<section id="optional-parameters">
<h2>Optional parameters<a class="headerlink" href="#optional-parameters" title="Permalink to this heading">¶</a></h2>
</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Princeton University.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>