# Analyze different type of data

The default values in the complete pipeline of HATCHet are typically used for analyzing whole-genome sequencing (WGS) data. However, when considering different type of data, as those from whole-exome sequencing (WES) data, users should adjust some of the parameters due to the different features of this kind of data. More specifically, there are 4 main points to consider when analyzing WES data:

  - *Larger bin sizes*. While a size of 50kb is standard for CNA analysis when considering whole-genome sequencing (WGS) data, data from whole-exome sequencing (WES) generally require to use large bin sizes in order to guarantee that each bin contains a sufficient number of heterozygous germline SNPs. Indeed, having a sufficient number of germline SNPs is needed to have good estimations with low variances for RDR and, especially, for the B-allele frequency (BAF) of each bin. As such, more appropriate bin sizes to consider may be 200kb or 250k when analyzing WES data; even larger bin sizes, e.g. `500kb`, may be needed for noisy WES data. However, one can use the informative plots from [plot-bins](https://github.com/raphael-group/hatchet/blob/master/doc/doc_plot_bins.html) to test different bin sizes and obtain the smallest size that allows low-variance estimates of RDR and BAF.
  
  - *Read-count thresholds*. As suggested in the GATK best practices, `count-alleles` requires two parameters -c (the minimum coverage for SNPs) and -C (the maximum coverage for SNPs) to reliably call SNPs and exclude those in regions with artifacts. GATK suggests to consider a value of -C that is at least twice larger than the average coverage and -c should be large enough to exclude non-sequenced regions. For example, `-c 6` and `-C 300` are values previously used for WGS data whose coverage is typically between 30x and 90x. However, WES data are generally characterized by a much larger average coverage and thus require larger values, e.g. `-c 20` and `-C 600`. These values are also very usefule to discard off-target regions. In any case, the user should ideally pick values according to the considered data.
  
  - *Bootstrapping for clustering*. Occasionally, WES may have very few points and much less data points than WGS. Only in these special cases with very few data points, the global clustering of cluster-bins may generally benefit from the integrated bootstrapping approach. This approach allow to generate a certain number of synthetic bins from the real ones to increase the power of the clustering. For example, the following cluster-bins parameters `-u 20 -dR 0.002 -dB 0.002` allow to activate the bootstraping which introduces 20 synthetic bins for each real bin with low variances.
  
  - *Genomic regions*. Provide a BED file defining the sequenced genomic regions when this is available. This helps to speed up the process.
